bittensor/_axon/__init__.py

```diff
@@ -180,4 +185,62 @@ def check_forward_callback( forward_callback:Callable, modality:int, pubkey:str
             
         if modality == bittensor.proto.Modality.TENSOR:
             sample_input = torch.rand(1,1,1)
-            forward_callback(pubkey,sample_input)
\ No newline at end of file
+            forward_callback(pubkey,sample_input)
+
+class AuthInterceptor(grpc.ServerInterceptor):
+    def __init__(self,keypair, key:str = 'Bittensor'):
+        r""" Creates a new server interceptor that authenticates incoming messages from passed arguments.
+        Args:
+            key (str, `optional`):
+                 key for authentication header in the metadata (default= Bittensor)
+            keypair (:obj:`callable`, `optional`): 
+                Substrate interface keypair object
+        """
+        self._valid_metadata = ('rpc-auth-header', key)
+        self.keypair = keypair
+        self.nounce_dic = {}
+
+        def deny(_, context):
+            context.abort(grpc.StatusCode.UNAUTHENTICATED, 'Invalid key')
+
+        self._deny = grpc.unary_unary_rpc_method_handler(deny)
+
+    def intercept_service(self, continuation, handler_call_details):
```

Comment:
Can we write this function with less nesting, I see 5 levels here which makes it hard to debug and understand

Reply:
yea, gonna rewrite parts of the logic, so it is easier to read and understand.